From b94779c2980218372e4e403f3ec25b75a861b134 Mon Sep 17 00:00:00 2001
From: Iaroslav Voitovych <yaroslav.voytovych@gmail.com>
Date: Wed, 7 Oct 2026 05:14:02 +0200
Subject: [PATCH BlueZ 2/2] unit/test-mgmt: Add tests for mgmt_unregister()
 from callbacks

Unregister from a notification callback the entry itself, the entry
after it, and an entry followed by mgmt_unregister_all(). Check that
an unregistered entry is not notified again, that its destroy callback
runs once and only after the notification, and that unregistering it
again fails. Also test unregistering outside a notification.
---
 unit/test-mgmt.c | 191 +++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 191 insertions(+)

diff --git a/unit/test-mgmt.c b/unit/test-mgmt.c
index 691af79a4..c0d58a0cd 100644
--- a/unit/test-mgmt.c
+++ b/unit/test-mgmt.c
@@ -416,6 +416,189 @@ static void test_destroy(gconstpointer data)
 	execute_context(context);
 }
 
+struct unregister_test;
+
+struct unregister_entry {
+	struct unregister_test *test;
+	unsigned int n;
+};
+
+struct unregister_test {
+	struct context *context;
+	unsigned int entries;
+	unsigned int events;
+	struct unregister_entry entry[3];
+	unsigned int id[3];
+	unsigned int called[3];
+	unsigned int destroyed[3];
+};
+
+static void unregister_destroy(void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+
+	g_assert_cmpuint(++entry->test->destroyed[entry->n], ==, 1);
+}
+
+static void unregister_add(struct unregister_test *test,
+				const struct command_test_data *data,
+				mgmt_notify_func_t callback)
+{
+	unsigned int n = test->entries++;
+
+	test->entry[n].test = test;
+	test->entry[n].n = n;
+	test->id[n] = mgmt_register(test->context->mgmt_client, data->opcode,
+					data->index, callback, &test->entry[n],
+					unregister_destroy);
+	g_assert_cmpuint(test->id[n], !=, 0);
+}
+
+static void unregister_from_cb(struct unregister_test *test, unsigned int n)
+{
+	struct mgmt *mgmt = test->context->mgmt_client;
+
+	g_assert_true(mgmt_unregister(mgmt, test->id[n]));
+	g_assert_false(mgmt_unregister(mgmt, test->id[n]));
+	g_assert_cmpuint(test->destroyed[n], ==, 0);
+
+	test->id[n] = 0;
+}
+
+static void count_cb(uint16_t index, uint16_t length, const void *param,
+							void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+
+	entry->test->called[entry->n]++;
+}
+
+static void unregister_self_cb(uint16_t index, uint16_t length,
+					const void *param, void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+
+	if (!entry->test->called[entry->n]++)
+		unregister_from_cb(entry->test, entry->n);
+}
+
+static void unregister_next_cb(uint16_t index, uint16_t length,
+					const void *param, void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+
+	if (!entry->test->called[entry->n]++)
+		unregister_from_cb(entry->test, entry->n + 1);
+}
+
+static void unregister_all_from_cb(uint16_t index, uint16_t length,
+					const void *param, void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+	struct unregister_test *test = entry->test;
+	struct mgmt *mgmt = test->context->mgmt_client;
+
+	test->called[entry->n]++;
+
+	unregister_from_cb(test, 1);
+	g_assert_true(mgmt_unregister_all(mgmt));
+	g_assert_false(mgmt_unregister(mgmt, test->id[2]));
+
+	context_quit(test->context);
+}
+
+static void unregister_quit_cb(uint16_t index, uint16_t length,
+					const void *param, void *user_data)
+{
+	struct unregister_entry *entry = user_data;
+	struct unregister_test *test = entry->test;
+	unsigned int i;
+
+	if (++test->called[entry->n] < test->events)
+		return;
+
+	/* Entries unregistered while notifying are freed by now */
+	for (i = 0; i < test->entries; i++)
+		g_assert_cmpuint(test->destroyed[i], ==, test->id[i] ? 0 : 1);
+
+	context_quit(test->context);
+}
+
+static void unregister_run(struct unregister_test *test,
+				const struct command_test_data *data,
+				const unsigned int *called)
+{
+	unsigned int i;
+
+	for (i = 0; i < test->events; i++)
+		g_assert_cmpint(write(test->context->fd, data->cmd_data,
+					data->cmd_size), ==, data->cmd_size);
+
+	execute_context(test->context);
+
+	for (i = 0; i < test->entries; i++) {
+		g_assert_cmpuint(test->called[i], ==, called[i]);
+		g_assert_cmpuint(test->destroyed[i], ==, 1);
+	}
+}
+
+static void test_unregister_self(gconstpointer data)
+{
+	struct unregister_test test = { .context = create_context(),
+								.events = 2 };
+	static const unsigned int called[] = { 1, 2, 2 };
+
+	unregister_add(&test, data, unregister_self_cb);
+	unregister_add(&test, data, count_cb);
+	unregister_add(&test, data, unregister_quit_cb);
+
+	unregister_run(&test, data, called);
+}
+
+static void test_unregister_next(gconstpointer data)
+{
+	struct unregister_test test = { .context = create_context(),
+								.events = 2 };
+	static const unsigned int called[] = { 2, 0, 2 };
+
+	unregister_add(&test, data, unregister_next_cb);
+	unregister_add(&test, data, count_cb);
+	unregister_add(&test, data, unregister_quit_cb);
+
+	unregister_run(&test, data, called);
+}
+
+static void test_unregister_id_all(gconstpointer data)
+{
+	struct unregister_test test = { .context = create_context(),
+								.events = 1 };
+	static const unsigned int called[] = { 1, 0, 0 };
+
+	unregister_add(&test, data, unregister_all_from_cb);
+	unregister_add(&test, data, count_cb);
+	unregister_add(&test, data, count_cb);
+
+	unregister_run(&test, data, called);
+}
+
+static void test_unregister_direct(gconstpointer data)
+{
+	struct unregister_test test = { .context = create_context(),
+								.events = 1 };
+	static const unsigned int called[] = { 0, 1 };
+	struct mgmt *mgmt = test.context->mgmt_client;
+
+	unregister_add(&test, data, count_cb);
+	unregister_add(&test, data, unregister_quit_cb);
+
+	g_assert_true(mgmt_unregister(mgmt, test.id[0]));
+	g_assert_cmpuint(test.destroyed[0], ==, 1);
+	g_assert_false(mgmt_unregister(mgmt, test.id[0]));
+	test.id[0] = 0;
+
+	unregister_run(&test, data, called);
+}
+
 int main(int argc, char *argv[])
 {
 	g_test_init(&argc, &argv, NULL);
@@ -435,6 +618,14 @@ int main(int argc, char *argv[])
 							test_unregister_all);
 	g_test_add_data_func("/mgmt/unregister/2", &event_test_1,
 							test_unregister_index);
+	g_test_add_data_func("/mgmt/unregister/3", &event_test_1,
+							test_unregister_self);
+	g_test_add_data_func("/mgmt/unregister/4", &event_test_1,
+							test_unregister_next);
+	g_test_add_data_func("/mgmt/unregister/5", &event_test_1,
+							test_unregister_id_all);
+	g_test_add_data_func("/mgmt/unregister/6", &event_test_1,
+							test_unregister_direct);
 
 	g_test_add_data_func("/mgmt/destroy/1", &event_test_1, test_destroy);
 
-- 
2.43.0

